File: 769.java
class Solution {
    public int maxChunksToSorted(int[] arr) {
        return find(arr, 0, arr.length, 0, new boolean[arr.length]);
    }

    private int find(int arr[], int i, int n, int s, boolean f[]) {
        if (i == n)
            return 0;
        int max = s;
        while (i < n) {
            f[arr[i]] = true;
            max = Math.max(max, arr[i]);
            if (allTrue(f, s, max))
                return 1 + find(arr, i + 1, n, max + 1, f);
            ++i;
        }
        return 1;
    }

    private boolean allTrue(boolean f[], int s, int e) {
        while (s <= e) {
            if (!f[s])
                return false;
            ++s;
        }
        return true;
    }
}

File: 769.py
class Solution:
    def allTrue(self, b: List[bool], s: int, e: int) -> bool:
        while s <= e:
            if not (b[s]):
                return False
            s += 1
        return True

    def find(self, arr: List[int], i: int, n: int, s: int, f: List[bool]) -> int:
        if i == n:
            return 0
        maxn = s
        while i < n:
            f[arr[i]] = True
            maxn = max(maxn, arr[i])
            if self.allTrue(f, s, maxn):
                return 1 + self.find(arr, i + 1, n, maxn + 1, f)
            i += 1
        return 1

    def maxChunksToSorted(self, arr: List[int]) -> int:
        n = len(arr)
        f = [False] * n
        return self.find(arr, 0, n, 0, f)
